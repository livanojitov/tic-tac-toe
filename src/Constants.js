const EMPTY = 0;
const COMPUTER = 1;
const USER = 2;
const TOTAL_SQUARES = 9;
const RANDOM = 1;
const HARD = 2;
const LOST = 0;
const WON = 1;
const DRAW = 2;
const UNKNOWN = 3;
const ENGLISH = 0;
const FRENCH = 1;
const SPANISH = 2;
const IMAGES_PER_CATEGORY = 20;

export {
  EMPTY,
  COMPUTER,
  USER,
  RANDOM,
  HARD,
  TOTAL_SQUARES,
  LOST,
  WON,
  DRAW,
  UNKNOWN,
  ENGLISH,
  FRENCH,
  SPANISH,
  IMAGES_PER_CATEGORY,
};
