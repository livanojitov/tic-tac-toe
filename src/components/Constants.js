const EMPTY = 0;

const COMPUTER = 1;
const USER = 2;

const EASY = 1;
const HARD = 2;
const NORMAL = 3;

const TOTAL_SQUARES = 9;

const CATEGORY = 'Category';

const YOU = 'You';
const OPPONENT = 'Computer';
const REFRESH = 'Refresh';
const USER_IMAGE_TOOLTIP = 'You will play with this image';
const COMPUTER_IMAGE_TOOLTIP = 'The computer will play with this image';
const REFRESH_TOOLTIP = 'Click to refresh the two images';

const START = 'Start';
const YES = 'Yes';
const NO = 'No';
const START_TOOLTIP = 'Who starts playing the game: You or the computer?';
const YES_TOOLTIP = 'You start playing the game.';
const NO_TOOLTIP = 'The computer starts playing the game.';
const GAME_STARTED_BY = 'Game started by:';

const LEVEL = 'Level';
const COMPUTER_PLAYS_EASY = 'Easy';
const COMPUTER_PLAYS_HARD = 'Hard';
const COMPUTER_PLAYS_NORMAL = 'Normal';
const LEVEL_TOOLTIP = 'How much thinking the computer does before playing.';
const COMPUTER_PLAYS_EASY_TOOLTIP = 'The computer plays randomly.';
const COMPUTER_PLAYS_HARD_TOOLTIP = 'The computer does some thinking before playing.';
const COMPUTER_PLAYS_NORMAL_TOOLTIP = 'The computer sometimes plays easy and sometimes plays hard.';
const LEVEL_HISTORY_EASY = 'Level: Easy';
const LEVEL_HISTORY_HARD = 'Level: Hard';
const LEVEL_HISTORY_NORMAL = 'Level: Normal';

const PLAY = 'Start the game';
const GAME_OVER = 'Game over';
const PLAY_AGAIN = 'Play again';
const LOST = 'You lost!';
const WON = 'You won!';
const DRAW = 'It\'s a draw!';

const GAME_WRITTEN_WITH = 'This game is written with:';
const SOURCE_CODE = 'Source Code:';
const OTHER_GAMES = 'Other games:';

const AUTHOR = 'Author:';
const EMAIL = 'Email:';
const GITHUB = 'Github:';
const LINKEDIN = 'LinkedIn:';
const HIRE_ME = 'Hire me!';

const NO_HISTORY_YET = 'No history yet. Play some games and comeback.';

const HOME = 'Home';
const HISTORY = 'History';
const ABOUT = 'About';
const CONTACT = 'Contact';

export { EMPTY, COMPUTER, USER, EASY, HARD, NORMAL, TOTAL_SQUARES, CATEGORY, YOU, OPPONENT, REFRESH, USER_IMAGE_TOOLTIP, COMPUTER_IMAGE_TOOLTIP, REFRESH_TOOLTIP, START, YES, NO, START_TOOLTIP, YES_TOOLTIP, NO_TOOLTIP, GAME_STARTED_BY, LEVEL, COMPUTER_PLAYS_EASY, COMPUTER_PLAYS_HARD, LEVEL_TOOLTIP, COMPUTER_PLAYS_EASY_TOOLTIP, COMPUTER_PLAYS_HARD_TOOLTIP, LEVEL_HISTORY_EASY, LEVEL_HISTORY_HARD, PLAY, GAME_OVER, PLAY_AGAIN, LOST, WON, DRAW, GAME_WRITTEN_WITH, SOURCE_CODE, OTHER_GAMES, AUTHOR, EMAIL, GITHUB, LINKEDIN, HIRE_ME, NO_HISTORY_YET, HOME, HISTORY, ABOUT, CONTACT, COMPUTER_PLAYS_NORMAL, COMPUTER_PLAYS_NORMAL_TOOLTIP, LEVEL_HISTORY_NORMAL };  
