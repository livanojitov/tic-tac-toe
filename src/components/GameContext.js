import React, { useState, createContext, useEffect } from 'react';
export const GameContext = createContext();

const GameContextProvider = (props) => {

  const [games, setGames] = useState([]);

  useEffect(() => {
    if (typeof(Storage) !== "undefined") {
      if (games.length){
        localStorage.tictactoe = JSON.stringify(games);
      }else{
        if (GameContextProvider.delete === 1){
          GameContextProvider.delete = 0;
          localStorage.removeItem('tictactoe');
        }else{
          if (localStorage.tictactoe){
            setGames(JSON.parse(localStorage.tictactoe));
          }
        }
      }
    }
  },[games]);

  const addGame = (game) => {
    game.id = Math.floor(Math.random() * 10000);
    setGames([...games, game]);
  };

  const removeGame = (id) => {
    if (games.length > 0){
      GameContextProvider.delete = 1;
      id = id * 1;
      setGames(games.filter(game => game.id !== id));   
    } 
  }

  const getGame = (id) => {
    if (games.length > 0){
      id = id * 1;
      return games.filter((game) => game.id === id)[0];
    }else{
      return false;
    }    
  }

  return (
    <GameContext.Provider value={{ games, addGame, removeGame, getGame }}>
      {props.children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;