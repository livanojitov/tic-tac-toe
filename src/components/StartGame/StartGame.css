.start{
  display: inline-block;
  width: 55%;  
}

.start select{
  margin-left: 2%;
}

.start span{
  cursor: default;
}