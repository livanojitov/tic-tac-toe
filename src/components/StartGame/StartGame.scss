.start {
  display: inline-block;
  width: 55%;

  select {
    margin-left: 2%;
  }

  span {
    cursor: default;
  }
}
