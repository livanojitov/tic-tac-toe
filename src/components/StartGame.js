import React, { Component } from 'react';
import * as constants from './Constants';
const { USER, START_TOOLTIP, START, YES_TOOLTIP, YES, COMPUTER, NO_TOOLTIP, NO, GAME_STARTED_BY, OPPONENT, YOU } = constants;

class StartGame extends Component {

  state = {
    who : USER
  } 

  render() {
    const { disabled, first, history } = this.props;
    const html = (typeof(history) === 'undefined'  || history === "false") ? (

      <div className="question">

        <span title = {START_TOOLTIP}>{START}?</span>

        <input onChange  = {this.setPlayer} 
                disabled = {disabled} 
                type     = "radio" 
                value    = {USER}
                name     = "player"
                title    = {YES_TOOLTIP}
                checked  = {this.state.who === USER}
                id       = "yes" />
        <label  disabled = {disabled} htmlFor = "yes" title = {YES_TOOLTIP}>{YES}</label>

        <input onChange  = {this.setPlayer}
                disabled = {disabled} 
                type     = "radio"
                value    = {COMPUTER}
                name     = "player"
                title    = {NO_TOOLTIP}
                checked  = {this.state.who === COMPUTER}
                id       = "no" />
        <label  disabled = {disabled} htmlFor = "no" title = {NO_TOOLTIP}>{NO}</label>   

      </div>
    ) : (
      <div className="question">
        {GAME_STARTED_BY} {first === COMPUTER ? OPPONENT : YOU}
      </div>
    );

    return (
      html
    )  
  }

  setPlayer = (e) => {
    const who = e.target.value * 1;
    this.setState(() => ({who}));
  }

  onPlayerChange = () => {
    if (this.props.onPlayerChange){
      this.props.onPlayerChange(this.state.who);
    }
  }

  componentDidUpdate(prevProps, prevState){
    if (prevState.who !== this.state.who){
      this.onPlayerChange();
    }     
  }

  componentDidMount(){
    this.onPlayerChange();
  }   
} 

export default StartGame;