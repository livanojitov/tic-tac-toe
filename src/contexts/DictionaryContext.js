import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";
import useLanguage from "../hooks/useLanguage";
import * as constants from "../Constants";

export const DictionaryContext = createContext();

const DictionaryContextProvider = ({ children }) => {
  let { data: dictionary, error } = useFetchData("dictionary");

  const [language, changeLanguageHandler] = useLanguage(constants.ENGLISH);

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  dictionary = dictionary[language];

  return (
    <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
