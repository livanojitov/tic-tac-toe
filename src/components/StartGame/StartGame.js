import React, { useContext, useMemo } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameContext } from "../../contexts/GameContext";
import { changeFirst } from "../../store/game/gameActions";
import { USER, COMPUTER } from "../../Constants";
import "./StartGame.scss";

const StartGame = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch } = useContext(GameContext);

  return useMemo(() => {
    return (
      <div className="start" data-testid="startgame">
        <span title={dictionary.START_TOOLTIP}>{dictionary.START}?</span>
        <select
          data-testid="select"
          defaultValue={game.first}
          onChange={(e) => {
            dispatch(changeFirst(e.target.value * 1));
          }}
          disabled={!game.disabled}
        >
          <option data-testid="select-option" key={0} value={USER}>
            {dictionary.YES}
          </option>
          <option data-testid="select-option" key={1} value={COMPUTER}>
            {dictionary.NO}
          </option>
        </select>
      </div>
    );
  }, [dictionary, game.first, game.disabled, dispatch]);
};

export default StartGame;
