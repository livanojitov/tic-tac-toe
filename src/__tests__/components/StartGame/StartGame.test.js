import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import { GameContext } from "../../../contexts/GameContext";
import StartGame from "../../../components/StartGame/StartGame";
import { USER } from "../../../Constants";

describe("StartGame", () => {
  let dictionary;
  let game;
  let dispatch;

  beforeEach(() => {
    dictionary = {
      START: "You Start",
      START_TOOLTIP: "Who starts playing the game: You or the computer?",
      YES: "Yes",
      NO: "No",
    };

    game = {
      first: USER,
      disabled: true,
    };

    dispatch = jest.fn();
  });

  const renderStartGame = () => {
    return render(
      <DictionaryContext.Provider value={{ dictionary }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <StartGame />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  it("renders correctly", () => {
    const { asFragment } = renderStartGame();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the start player dropdown", () => {
    renderStartGame();

    const startGame = screen.getByTestId("startgame");
    expect(startGame).not.toBeNull();
    expect(startGame).toHaveClass("start");
  });

  it("should change the starting player in the dropdown", () => {
    renderStartGame();

    expect(
      screen.getByRole("option", { name: dictionary.YES }).selected
    ).toBeTruthy();
    expect(
      screen.getByRole("option", { name: dictionary.NO }).selected
    ).toBeFalsy();

    userEvent.selectOptions(
      screen.getByTestId("select"),
      screen.getByText(dictionary.NO)
    );

    expect(
      screen.getByRole("option", { name: dictionary.YES }).selected
    ).toBeFalsy();
    expect(
      screen.getByRole("option", { name: dictionary.NO }).selected
    ).toBeTruthy();
  });
});
