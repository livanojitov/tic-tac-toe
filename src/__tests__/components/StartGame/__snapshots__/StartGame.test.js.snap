// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartGame renders correctly 1`] = `
<DocumentFragment>
  <div
    class="start"
    data-testid="startgame"
  >
    <span
      title="Who starts playing the game: You or the computer?"
    >
      You Start?
    </span>
    <select
      data-testid="select"
    >
      <option
        data-testid="select-option"
        selected=""
        value="2"
      >
        Yes
      </option>
      <option
        data-testid="select-option"
        value="1"
      >
        No
      </option>
    </select>
  </div>
</DocumentFragment>
`;
